typing/typecore.ml

```diff
@@ -3056,22 +3063,18 @@ and type_expect_
         | _ ->
             funct, sargs
       in
-      begin_def ();
       let (args, ty_res) = type_application env funct sargs in
-      end_def ();
-      unify_var env (newvar()) funct.exp_type;
```

Comment:
What is going on here? It looks like the behavior is changed.

This seems related to #11519 but the change is different from the one in #11519.

Unless you are very confident that the change is correct, I think we could stick to "just a refactoring" (no change in behavior) in this current PR to make it easier to review, and keep small behavior changes for smaller independent PRs.

Reply:
I had to scratch my head for a while to understand why this extra level was needed. It is actually there to warn on partial applications, so I moved the `with_local_level` to the end of `type_application`, where it is needed.
Note that `type_application` is a long sequence of functions, called from the wrapped code, so there is no semantical change.
I also removed the call to `unify_var`. as there is no need to lower the levels of `funct.exp_type`, since it does not appear in the return type. I don't know why I included this line in 2004, probably did not think much about it.